programs/bpf_loader/src/syscalls/mod.rs

```diff
@@ -993,6 +1002,52 @@ declare_syscall!(
     }
 );
 
+declare_syscall!(
+    /// keccak sponge permutation - not to be confused with the keccak256 hash
+    /// chances are you're looking for `sol_keccak256` instead
+    /// do not use this unless you know what you're doing
+    SyscallKeccakPermutation,
+    fn call(
+        &mut self,
+        state_addr: u64,
+        _arg2: u64,
+        _arg3: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &mut MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        let invoke_context = question_mark!(
+            self.invoke_context
+                .try_borrow()
+                .map_err(|_| SyscallError::InvokeContextBorrowFailed),
+            result
+        );
+        let compute_budget = invoke_context.get_compute_budget();
+        question_mark!(
+            invoke_context
+                .get_compute_meter()
+                .consume(compute_budget.keccak_permutation_cost),
```

Comment:
3 CU seems a bit low. Other syscalls are charging 85 CU base cost and ~10 CU for the memory access

Reply:
ok - is there a rule for how these costs are determined? I'm not sure the "base costs + memory access" makes sense here because it's a fixed-sized input.

Reply:
They are priced to approximate real-world execution time. For syscalls with static complexity, other code uses 100 CU minimum, for example. I can do some fuzzing & micro-benchmarks if it helps.

https://github.com/solana-labs/solana/blob/d9c7bc7e780b4c6ef97d3b75263048f1b8825131/program-runtime/src/compute_budget.rs#L60-L61

Reply:
I see, thanks. We can use that one.

Reply:
in terms of ns? us? If you can that would help a lot!

Reply:
in terms of ns? us?

Reply:
Having learned this, I think I'll circle back with a different version that operates on variable-length input with cost "85 CU base + 10 CU / block"